### PR TITLE
fix #25 Allow listen until LoopResource/PoolResources disposal

### LINKED ISSUE #25: No way to wait for shutdown of TcpClient resources
Following the model of `TcpResources` I'm trying to create my own resources:
````java
this.channelGroup = new DefaultChannelGroup(ImmediateEventExecutor.INSTANCE);
this.loopResources = LoopResources.create("reactor-netty-tcp-client");
this.poolResources = PoolResources.fixed("reactor-netty-tcp-pool");

TcpClient.create(opts -> opts
		.channelGroup(this.channelGroup)
		.loopResources(this.loopResources)
		.poolResources(this.poolResources)
		.preferNative(false));
````
So I can then shut them down:
````java
ChannelGroupFuture future = this.channelGroup.close();
Mono<Void> completion = FutureMono.from(future)
		.doAfterTerminate((x, e) -> {
			this.loopResources.dispose();
			this.poolResources.dispose();
		});
````
Note however that unlike `ChannelGroup` there is no option to wait for the completion of the shutdown of resources whose dispose methods return void and don't wait.

### PR DESCRIPTION
**Pull request by simonbasle**
This commit adds a disposeAndListen method to both LoopResources and
PoolResources that allow to dispose these but also get a notification of
when the disposal is done, in a best effort fashion.

Subsequent calls however are not guaranteed to return a relevant Mono,
which could complete sooner than the actual disposal as triggered by the
first invocation.

### CODE REVIEW

File: src/main/java/reactor/ipc/netty/resources/DefaultLoopResources.java [hunk 1]
```diff
@@ -85,24 +88,44 @@ static ThreadFactory threadFactory(DefaultLoopResources parent, String prefix) {
 
 	@Override
 	public void dispose() {
+		disposeAndListen();
+	}
+
+	public Mono<Void> disposeAndListen() {
```
Comment (smaldini) [inline src/main/java/reactor/ipc/netty/resources/DefaultLoopResources.java]: use deferDispose() and then Mono.defer inside to defer the actual dispose

File: src/main/java/reactor/ipc/netty/resources/DefaultPoolResources.java [hunk 2]
```diff
@@ -188,19 +188,22 @@ public String toString() {
...
+			Pool pool;
+			for (SocketAddress key: channelPools.keySet()) {
+				pool = channelPools.remove(key);
+				if(pool != null){
+					pool.close();
+				}
 			}
-		}
-		return Mono.empty();
+			return Mono.empty();
```
Comment (smaldini) [inline src/main/java/reactor/ipc/netty/resources/DefaultPoolResources.java]: Could be Mono.fromRunnable then

File: src/main/java/reactor/ipc/netty/resources/DefaultLoopResources.java [hunk 3]
```diff
@@ -87,45 +87,43 @@ static ThreadFactory threadFactory(DefaultLoopResources parent, String prefix) {
...
-			group = cacheNativeServerLoops.get();
-			if(group != null){
-				Future<?> cnsrvlFuture = group.shutdownGracefully();
-				cnsrvlMono = FutureMono.from((Future)cnsrvlFuture);
-			}
-
-			return Mono.when(clMono, slMono, cnclMono, cnslMono, cnsrvlMono).then();
-		}
-		return Mono.empty();
+			return Mono.empty();
```
Comment (smaldini) [inline src/main/java/reactor/ipc/netty/resources/DefaultLoopResources.java]: could be Mono.fromRunnable

**Comment by smaldini** [comment]
Would deferDispose() works better than disposeDeferred ?

**Comment by smaldini** [comment]
Use shutdownLater ?

**Comment by simonbasle** [comment]
perfect 👌

**Pull request merged**